src/crystal/system/unix/evented/event_loop.cr

```diff
@@ -0,0 +1,517 @@
+      end
+
+      pd.value.remove(io.fd)
+    end
+  end
+
+  private def internal_remove(io)
+    return unless (index = io.__evloop_data).valid?
+    io.__evloop_data = Arena::INVALID_INDEX
+
+    Evented.arena.free(index) do |pd|
+      pd.value.remove(io.fd) { } # ignore system error
+    end
+  end
+
+  private def wait_readable(io, timeout = nil) : Nil
+    wait_readable(io, timeout) do
+      raise IO::TimeoutError.new("Read timed out")
+    end
+  end
+
+  private def wait_writable(io, timeout = nil) : Nil
+    wait_writable(io, timeout) do
+      raise IO::TimeoutError.new("Write timed out")
+    end
+  end
+
+  private def wait_readable(io, timeout = nil, &) : Nil
+    yield if wait(:io_read, io, timeout) do |pd, event|
+               # don't wait if the waiter has already been marked ready (see Waiters#add)
+               return unless pd.value.@readers.add(event)
+             end
+  end
+
+  private def wait_writable(io, timeout = nil, &) : Nil
+    yield if wait(:io_write, io, timeout) do |pd, event|
+               # don't wait if the waiter has already been marked ready (see Waiters#add)
+               return unless pd.value.@writers.add(event)
+             end
+  end
+
+  private def wait(type : Evented::Event::Type, io, timeout, &)
+    # get or allocate the poll descriptor
+    if (index = io.__evloop_data).valid?
+      pd = Evented.arena.get(index)
+    else
+      pd, index = Evented.arena.lazy_allocate(io.fd) do |pd, index|
+        # register the fd with the event loop (once), it should usually merely add
+        # the fd to the current evloop but may "transfer" the ownership from
+        # another event loop:
+        io.__evloop_data = index
+        pd.value.take_ownership(self, io.fd, index)
+      end
+    end
+
+    # create an event (on the stack)
+    event = Evented::Event.new(type, Fiber.current, index, timeout)
+
+    # add the event to the waiting list
+    yield pd, pointerof(event)
+
+    if event.wake_at?
+      add_timer(pointerof(event))
+
+      Fiber.suspend
+
+      # no need to delete the timer: either it triggered which means it was
+      # dequeued, or `#unsafe_resume_io` was called to resume the IO and the
+      # timer got deleted from the timers before the fiber got reenqueued.
+      return event.timed_out?
+    end
+
+    Fiber.suspend
+    false
+  end
+
+  private def check_open(io : IO)
+    raise IO::Error.new("Closed stream") if io.closed?
+  end
+
+  # internals: timers
+
+  protected def add_timer(event : Evented::Event*)
+    @lock.sync do
+      is_next_ready = @timers.add(event)
+      system_set_timer(event.value.wake_at) if is_next_ready
+    end
+  end
+
+  protected def delete_timer(event : Evented::Event*) : Bool
+    @lock.sync do
+      dequeued, was_next_ready = @timers.delete(event)
+      if dequeued
+        # update system timer if we deleted the next timer
+        system_set_timer(@timers.next_ready?) if was_next_ready
+        return true if dequeued
```

Comment:
```suggestion
        return true
```

Reply:
Thanks, we can even just return `dequeued`.